hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinySessionFactoryImpl.java

```diff
@@ -73,16 +73,33 @@ public MutinySessionFactoryImpl(SessionFactoryImpl delegate) {
 	Uni<Mutiny.Session> newSession() throws HibernateException {
 		SessionCreationOptions options = options();
 		return uni( () -> connection( options.getTenantIdentifier() ) )
-				.map( reactiveConnection -> new ReactiveSessionImpl( delegate, options, reactiveConnection ) )
+				.chain( reactiveConnection -> create( reactiveConnection, () -> new ReactiveSessionImpl( delegate, options, reactiveConnection ) ) )
 				.map( s -> new MutinySessionImpl(s, this) );
 	}
 
 	Uni<Mutiny.Session> newSession(String tenantId) throws HibernateException {
 		return uni( () -> connection( tenantId ) )
-				.map( reactiveConnection -> new ReactiveSessionImpl( delegate, options( tenantId ), reactiveConnection ) )
+				.chain( reactiveConnection -> create( reactiveConnection, () -> new ReactiveSessionImpl( delegate, options( tenantId ), reactiveConnection ) ) )
 				.map( s -> new MutinySessionImpl(s, this) );
 	}
 
+	/**
+	 * Close the connection if something goes wrong during the creation of the session
+	 */
+	private <S> Uni<S> create(ReactiveConnection connection, Supplier<S> supplier) {
+		try {
+			return Uni.createFrom().item( supplier.get() );
```

Comment:
@DavideD this looks weird to me. Did you mean to write `Uni.createFrom().item(supplier)`?

Reply:
Oh no, you didn't, because you want to force the error to happen immediately.

But then why wrap it in a `Uni`?

Reply:
Oh I see the point is to convert a `CompletionStage` to a `Uni`. But is that even necessary here?

Hard to make sense of this change from the diff.

Reply:
@DavideD why didn't you just use `onFailure()` to catch the exception and close the session. Wouldn't make the code a little easier to understand?

Reply:
Sorry, I'm still working a bit around it. I've already applied your suggestions.